coderd/workspaceapps_test.go

```diff
@@ -165,6 +230,164 @@ func TestWorkspaceAppsProxyPath(t *testing.T) {
 		resp, err := client.Request(ctx, http.MethodGet, "/@me/"+workspace.Name+"/apps/fake/", nil)
 		require.NoError(t, err)
 		defer resp.Body.Close()
+		// this is 200 OK because it returns a dashboard page
+		require.Equal(t, http.StatusOK, resp.StatusCode)
+	})
+}
+
+func TestWorkspaceAppsProxySubdomain(t *testing.T) {
+	t.Parallel()
+	client, orgID, workspace, port := setupProxyTest(t)
+
+	// proxyURL generates a URL for the proxy subdomain. The default path is a
+	// slash.
+	proxyURL := func(t *testing.T, appNameOrPort interface{}, pathAndQuery ...string) string {
+		t.Helper()
+
+		var (
+			appName string
+			port    uint16
+		)
+		if val, ok := appNameOrPort.(string); ok {
+			appName = val
+		} else {
+			port, ok = appNameOrPort.(uint16)
+			require.True(t, ok)
+		}
+
+		me, err := client.User(context.Background(), codersdk.Me)
+		require.NoError(t, err, "get current user details")
+
+		hostname := httpapi.ApplicationURL{
+			AppName:       appName,
+			Port:          port,
+			AgentName:     proxyTestAgentName,
+			WorkspaceName: workspace.Name,
+			Username:      me.Username,
+			BaseHostname:  "test.coder.com",
+		}.String()
+
+		actualPath := "/"
+		query := ""
+		if len(pathAndQuery) > 0 {
+			actualPath = pathAndQuery[0]
+		}
+		if len(pathAndQuery) > 1 {
+			query = pathAndQuery[1]
+		}
+
+		return (&url.URL{
+			Scheme:   "http",
+			Host:     hostname,
+			Path:     actualPath,
+			RawQuery: query,
+		}).String()
+	}
+
+	t.Run("NoAuthShould401", func(t *testing.T) {
```

Comment:
Could we redirect to `/login` with a `redirect` query parameter back to the subdomain URL instead? Seems like this would be a questionable UX.

Reply:
I've made it redirect to login but it's impossible to redirect back to the subdomain because we can't determine whether the accessed scheme was HTTP or HTTPS. If users are using a reverse proxy in front of coder (which they likely will) then coder will think it's receiving HTTP traffic when the actual scheme is HTTPS.